cpu/esp32/periph/gpio.c

```diff
@@ -537,13 +541,110 @@ int8_t gpio_is_rtcio (gpio_t pin)
     return _gpio_to_rtc[pin];
 }
 
-int gpio_config_sleep_mode (gpio_t pin, bool mode, bool input)
-{
-    return rtcio_config_sleep_mode (pin, mode, input);
-}
-
 int gpio_set_direction(gpio_t pin, gpio_mode_t mode)
 {
     /* TODO implementation, for the moment we simply initialize the GPIO */
     return gpio_init(pin, mode);
 }
+
+int gpio_set_drive_capability(gpio_t pin, gpio_drive_strength_t drive)
+{
+    assert(pin < GPIO_PIN_NUMOF);
+    assert(pin < GPIO34);
+
+    const struct _rtc_gpio_t* rtc = (_gpio_to_rtc[pin] != -1) ?
+                                     &_rtc_gpios[_gpio_to_rtc[pin]] : NULL;
+
+    SET_PERI_REG_BITS(_gpio_to_iomux_reg, FUN_DRV_V, drive, FUN_DRV_S);
+    if (rtc) {
+        SET_PERI_REG_BITS(rtc->reg, 0x3, drive, rtc->drive);
+    }
+    return 0;
+}
+
+#if MODULE_PERIPH_GPIO_IRQ
+static uint32_t gpio_int_saved_type[GPIO_PIN_NUMOF];
+#endif
+
+void gpio_pm_sleep_enter(unsigned mode)
+{
+    /*
+     * Activate the power domain for RTC peripherals either when
+     * ESP_PM_GPIO_HOLD is defined or when light sleep mode is activated.
+     * As long as the RTC peripherals are active, the pad state of RTC GPIOs
+     * is held in deep sleep and the pad state of all GPIOs is held in light
+     * sleep.
+     */
+#ifdef ESP_PM_GPIO_HOLD
+    esp_sleep_pd_config(ESP_PD_DOMAIN_RTC_PERIPH, ESP_PD_OPTION_ON);
+#else
+    if (mode == ESP_PM_LIGHT_SLEEP) {
+        esp_sleep_pd_config(ESP_PD_DOMAIN_RTC_PERIPH, ESP_PD_OPTION_ON);
+    }
+#endif
+
+    if (mode == ESP_PM_DEEP_SLEEP) {
+#ifdef ESP_PM_WUP_PINS
+        static const gpio_t wup_pins[] = { ESP_PM_WUP_PINS };
+        /*
+         * Prepare the wake-up pins if a single pin or a comma-separated list of
+         * pins is defined for wake-up.
+         */
+        uint64_t wup_pin_mask = 0;
```

Comment:
So there are less than 32 ADC pins - `uint32_t` would suffice here.

Reply:
No, the SDK requires `uint64_t`
```
esp_err_t esp_sleep_enable_ext1_wakeup(uint64_t mask, esp_sleep_ext1_wakeup_mode_t mode);
```
because GPIOs are numbered from 0 to 39 even though only 18 of them are RTC GPIOs and can be used as wake-up source.